File: main/smallest-range/smallest-range.py
import contextlib

class AndCounter:
    '''Tracks frequencies of items and knows if it has at least one of each.'''
    def __init__(self, size):
        self._counts = [0] * size
        self._missing = size

    def __bool__(self):
        return not self._missing

    def add(self, item):
        if not self._counts[item]:
            self._missing -= 1
        self._counts[item] += 1

    def remove(self, item):
        self._counts[item] -= 1
        if not self._counts[item]:
            self._missing += 1

class Solution:
    @staticmethod
    def smallestRange(nums):
        """
        :type nums: List[List[int]]
        :rtype: List[int]
        """
        pairs = Solution._markChainAndSort(nums)
        return Solution._findMinRangeOfFirstCoveringAllSecond(pairs, len(nums))

    @staticmethod
    def _markElementsByRow(nums):
        for i, row in enumerate(nums):
            fixed_i = i # so later iterations don't affect "earlier" results
            yield ((elem, fixed_i) for elem in row)

    @staticmethod
    def _markChainAndSort(nums):
        groups = Solution._markElementsByRow(nums)
        chain = itertools.chain.from_iterable(groups)
        return sorted(chain, key=operator.itemgetter(0))

    @staticmethod
    def _findMinRangeOfFirstCoveringAllSecond(pairs, second_population):
        min_first_range = range(pairs[0][0], pairs[-1][0])
        second_targets = AndCounter(second_population)
        right_iter = iter(pairs)

        with contextlib.suppress(StopIteration):
            for left_first, left_second in pairs:
                while not second_targets:
                    right_first, right_second = next(right_iter)
                    second_targets.add(right_second)

                if right_first - left_first < len(min_first_range):
                    min_first_range = range(left_first, right_first)

                second_targets.remove(left_second)

        return min_first_range.start, min_first_range.stop

File: main/smallest-range/smallest-range.java
import java.util.ConcurrentModificationException;

class Solution {
    public int[] smallestRange(final List<List<Integer>> nums) {
        final Element[] elems = flatten(nums);
        Arrays.sort(elems);
        return findMinRangeCoveringAllGroups(elems, nums.size());
    }

    private static Element[] flatten(final List<List<Integer>> nums) {
        final Element[] elems = new Element[getTotalLength(nums)];

        int k = 0;
        for (int rowId = 0; rowId != nums.size(); ++rowId) {
            for (final int value : nums.get(rowId))
                elems[k++] = new Element(value, rowId);
        }

        if (k != elems.length) throw new ConcurrentModificationException();
        return elems;
    }

    private static <T> int getTotalLength(final List<List<T>> rows) {
        int len = 0;
        for (final List<T> row : rows) len += row.size();
        return len;
    }

    private static int[] findMinRangeCoveringAllGroups(final Element[] elems,
                                                       final int rowCount) {
        int start = elems[0].value, stop = elems[elems.length - 1].value;
        final AndCounter targets = new AndCounter(rowCount);

        for (int left = -1, right = -1; ; ) {
            while (!targets.atGoal() && ++right != elems.length)
                targets.add(elems[right].rowId);

            if (right == elems.length) break;

            while (targets.atGoal())
                targets.remove(elems[++left].rowId);

            if (elems[right].value - elems[left].value < stop - start) {
                start = elems[left].value;
                stop = elems[right].value;
            }
        }

        return new int[] { start, stop };
    }
}

/** A value that remembers what row it came from. */
final class Element implements Comparable<Element> {
    Element(final int value, final int rowId) {
        this.value = value;
        this.rowId = rowId;
    }

    public int compareTo(final Element other) {
        return Integer.compare(value, other.value);
    }

    final int value, rowId;
}

/** Tracks frequencies of items and knows if it has at least one of each. */
final class AndCounter {
    AndCounter(final int size) {
        _counts = new int[size];
        _missing = size;
    }

    void add(final int item) {
        if (_counts[item]++ == 0) --_missing;
    }

    void remove(final int item) {
        if (--_counts[item] == 0) ++_missing;
    }

    boolean atGoal() {
        return _missing == 0;
    }

    private final int[] _counts;

    private int _missing;
}
